subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -110,22 +145,56 @@ private boolean experimentalCompilationAvoidanceEnabled() {
     }
 
     @Override
-    @TaskAction
     protected void compile() {
+        compile(null);
+    }
+
+    @TaskAction
+    protected void compile(InputChanges inputChanges) {
         checkGroovyClasspathIsNonEmpty();
         warnIfCompileAvoidanceEnabled();
+
+        if (!getOptions().isIncremental() || compilerPluginClasspathChanged(inputChanges)) {
+            doCompile(null, null);
+        } else {
+            Multimap<File, String> oldMappings = readSourceClassesMappingFile(sourceClassesMappingFile);
+            doCompile(inputChanges, oldMappings);
+            updateSourceClassesMappingFile(inputChanges, oldMappings);
+        }
+    }
+
+    private boolean compilerPluginClasspathChanged(InputChanges inputChanges) {
+        return experimentalCompilationAvoidanceEnabled() && inputChanges.getFileChanges(getCompilerPluginClasspath()).iterator().hasNext();
+    }
+
+    private void doCompile(InputChanges inputChanges, Multimap<File, String> sourceClassesMapping) {
         DefaultGroovyJavaJointCompileSpec spec = createSpec();
-        WorkResult result = getCompiler(spec).execute(spec);
+        WorkResult result = getCompiler(spec, inputChanges, sourceClassesMapping).execute(spec);
         setDidWork(result.getDidWork());
     }
 
+    private void updateSourceClassesMappingFile(InputChanges inputChanges, Multimap<File, String> oldMappings) {
+        Multimap<File, String> mappingsDuringIncrementalCompilation = readSourceClassesMappingFile(sourceClassesMappingFile);
+
+        for (FileChange fileChange : inputChanges.getFileChanges(getStableSources())) {
+            File changedFile = fileChange.getFile();
+            if (fileChange.getChangeType() == ChangeType.REMOVED) {
+                oldMappings.removeAll(changedFile);
+            } else {
+                // Is it possible to get empty result from mappingsDuringIncrementalCompilation.get(changedFile) ?
+                // I.e. a source file doesn't generate any classes.
+                oldMappings.putAll(changedFile, mappingsDuringIncrementalCompilation.get(changedFile));
+            }
+        }
+    }
+
     private void warnIfCompileAvoidanceEnabled() {
         if (experimentalCompilationAvoidanceEnabled()) {
             DeprecationLogger.incubatingFeatureUsed("Groovy compilation avoidance");
         }
     }
 
-    private Compiler<GroovyJavaJointCompileSpec> getCompiler(GroovyJavaJointCompileSpec spec) {
+    private Compiler<GroovyJavaJointCompileSpec> getCompiler(GroovyJavaJointCompileSpec spec, InputChanges inputChanges, Multimap<File, String> sourceClassesMapping) {
         if (compiler == null) {
```

Comment:
Can we remove the `compiler` field? This method is only called once anyway.

Reply:
Removed.